sbi/inference/snle/mnle.py

```diff
@@ -0,0 +1,239 @@
+                `.log_prob` and `.sample()`.
+            device: Training device, e.g., "cpu", "cuda" or "cuda:{0, 1, ...}".
+            logging_level: Minimum severity of messages to log. One of the strings
+                INFO, WARNING, DEBUG, ERROR and CRITICAL.
+            summary_writer: A tensorboard `SummaryWriter` to control, among others, log
+                file location (default is `<current working directory>/logs`.)
+            show_progress_bars: Whether to show a progressbar during simulation and
+                sampling.
+            unused_args: Absorbs additional arguments. No entries will be used. If it
+                is not empty, we warn. In future versions, when the new interface of
+                0.14.0 is more mature, we will remove this argument.
+        """
+
+        if isinstance(density_estimator, str):
+            assert (
+                density_estimator == "mnle"
+            ), f"""MNLE can be used with preconfigured 'mnle' density estimator only, 
+                not with {density_estimator}."""
+        kwargs = del_entries(locals(), entries=("self", "__class__", "unused_args"))
+        super().__init__(**kwargs, **unused_args)
+
+    def append_simulations(
+        self,
+        theta: Tensor,
+        x: Tensor,
+        from_round: int = 0,
+    ) -> "LikelihoodEstimator":
+        r"""Store parameters and simulation outputs to use them for later training.
+
+        Data are stored as entries in lists for each type of variable (parameter/data).
+
+        Stores $\theta$, $x$, prior_masks (indicating if simulations are coming from the
+        prior or not) and an index indicating which round the batch of simulations came
+        from.
+
+        Args:
+            theta: Parameter sets.
+            x: Simulation outputs.
+            from_round: Which round the data stemmed from. Round 0 means from the prior.
+                With default settings, this is not used at all for `SNLE`. Only when
+                the user later on requests `.train(discard_prior_samples=True)`, we
+                use these indices to find which training data stemmed from the prior.
+
+        Returns:
+            NeuralInference object (returned so that this function is chainable).
+        """
+
+        theta, x = validate_theta_and_x(theta, x, training_device=self._device)
+        # Test net here to avoid copying the train function from base.
+        assert isinstance(
+            self._build_neural_net(theta, x), MixedDensityEstimator
+        ), """Invalid density estimtor for MNLE, pass 'mnle' or pass a built function
+        similar to sbi.neural_nets.mnle.build_mnle"""
+
+        self._theta_roundwise.append(theta)
+        self._x_roundwise.append(x)
+        self._prior_masks.append(mask_sims_from_prior(int(from_round), theta.size(0)))
+        self._data_round_index.append(int(from_round))
+
+        return self
+
+    def train(
```

Comment:
Is this an exact copy of `super.train()`? If yes, why do we keep it here at all?

Reply:
I think it needs to delete the two entries in here, just like for `snle_a`, no?:
```python
        kwargs = del_entries(locals(), entries=("self", "__class__"))
        return super().train(**kwargs)
```

Reply:
hm, I'm not sure. I think we can delete the `train()` method also for `SNLE_A()`. The deletion of these two arguments is only because we pass on all `locals()`. So I think it would be no problem to just delete the entire `train()` method in the child classes of SNLE.

Reply:
Ok I will test that.

Reply:
interesting: when I remove the child `train` then the tests take 10x longer. so this induces a large speed up. any ideas why?

Reply:
Super weird. Absolutely no idea. Maybe just bad luck with MCMC? But really just guessing

Reply:
indeed, I repeated the comparison and it seems this was just by chance. I will remove the train methods then.